man/tpm2_setprimarypolicy.1.md
% tpm2_setprimarypolicy(1) tpm2-tools | General Commands Manual

# NAME

**tpm2_setprimarypolicy**(1) - Sets the authorization policy for the lockout
(lockoutPolicy), the platform hierarchy (platformPolicy), the storage hierarchy
(ownerPolicy), and the endorsement hierarchy (endorsementPolicy).

# SYNOPSIS

**tpm2_setprimarypolicy** [*OPTIONS*]

# DESCRIPTION

**tpm2_setprimarypolicy**(1) -  Sets the authorization policy for the lockout
(lockoutPolicy), the platform hierarchy (platformPolicy), the storage hierarchy
(ownerPolicy), and the endorsement hierarchy (endorsementPolicy).

# OPTIONS

These options control creating the policy authorization session:

  * **-C**, **\--hierarchy**=_OBJECT_:

    Specifies the hierarchy whose authorization policy is to be setup. It can be
    specified as o|p|e|l

 * **-P**, **\--auth**=_AUTH_:

    Specifies the authorization value for the hierarchy.

  * **-L**, **\--policy**=_FILE_:

    The file path of the authorization policy data.

  * **-g**, **\--hash-algorithm**=_ALGORITHM_:

    The hash algorithm used in computation of the policy digest.

 * **\--cphash**=_FILE_

    File path to record the hash of the command parameters. This is commonly
    termed as cpHash. NOTE: When this option is selected, The tool will not
    actually execute the command, it simply returns a cpHash.

## References

[context object format](common/ctxobj.md) details the methods for specifying
_OBJECT_.

[authorization formatting](common/authorizations.md) details the methods for
specifying _AUTH_.

[algorithm specifiers](common/alg.md) details the options for specifying
cryptographic algorithms _ALGORITHM_.

[common options](common/options.md) collection of common options that provide
information many users may expect.

[common tcti options](common/tcti.md) collection of options used to configure
the various known TCTI modules.


# EXAMPLES

## Set a blank authorization policy for endorsement hierarchy
```bash
tpm2_setprimarypolicy -C e
```

[returns](common/returns.md)

[footer](common/footer.md)


CHANGELOG
v2.2.2 - next:
*  fix make distcheck
v2.1.1 - next:
* Potential memory leak fix when tcti/sapi initialization fails.
* tpm2_listpcrs: use TPM2_GetCapability to determine PCRs to read
* listpcrs: remove one redundant call to tpm get cap
* listpcrs: fix for unsupported/disabled alg in -L
* build: use supported comment to suppress GCC7 fallthrough warning
* kdfa: allow to build with OpenSSL 1.1.x

v2.1.0 - 2017-7-24
* Fix readx and writex on multiple EINTR returns.
* Add support for the tabrmd TCTI. This is the new default.
* Change default socket port from 2323 (the old resourcemgr) to 2321 (default
  simulator port).
* Cherry-pick fix for CVE-2017-7524.
* Fix tpm2_listpcr command line option handling.
* Fix tpm2_getmanufec memory issues.

v2.0 - 2017-03-29
* Tracked on the milestone: https://github.com/01org/tpm2.0-tools/milestone/2
* Reworked all the tools to support configurable TCTIs, based on build time
  configuration, one can specify the tcti via the --tcti (-T) option to all
  tools.
* tpm2_getrandom interface made -s a positional argument.
* Numerous bug fixes.

v1.1 - 2016-11-04
Added
* travis ci support.
* Allow for unit tests to be enabled selectively.
* tpm2_rc_decode tool: Decode TPM_RC error codes.
* Android Make file
* tpm2_listpersistent: list all persistent objects
* test scripts for tpm2-tools
* tpm2_nvreadlock
* tpm2_getmanufec: retrieve EC from tpm manufacturer server.
* Copy 'common' and 'sample' code from the TPM2.0-TSS repo.

Modified
* tpm2_takeownership: update option -c to use lockout password to clear.
* tpm2_listpcrs: add options -L and -s, rewrite to increase performance.
* tpm2_quote: added -L option to support selection of multiple banks.
* tpm2_quote: add -q option to get qualifying data.
* configure: Use pkg-config to get info about libcurl and libcrypto.
* configure: Use pkg-config to locate SAPI and TCTI headers / libraries.
* tpm2_x: Add -X option to enable password input in Hex format.
* tpm2_nvdefine: Change -X option to -I.
* tpm2-nvwrite: fix for unable to write 1024B+ data.
* tpm2_getmanufec: Fix base64 encoding.
* tpm2_x: fixed a lot of TPM2B failures caused by wrong initialization.
* tpm2_getmanufec: let configure handle libs.
* tpm2_getmanufec: Convert from dos to unix format.
* build: Check for TSS2 library @ configure time.
* build: Detect required TSS2 and TCTI headers.
* build: Use libtool to build the common library
* build: Install all binaries into sbin.
* build: Build common sources into library.
* build: Move all source files to 'src'.
* Makefile.am: Move all build rules into single Makefile.am.
* everything: Use new TCTI headers and fixup API calls.
* everything: Update source to cope with sapi header cleanup.
* tpm2_activatecredential: Updated to support TCG compatible EK
* tpm2_getpubak: Updated to use TCG compatible EK
* tpm2_getpubek: fix ek creation to follow TCG EK profile spec.

Removed
* Windows related code
* depenedency on the TPM2.0-TSS repo source code

v1.0 - 2015-10-19
* 1.0 release
* 29 tools included


tools/tpm2_setprimarypolicy.c
/* SPDX-License-Identifier: BSD-3-Clause */
#include <stdlib.h>

#include "files.h"
#include "log.h"
#include "tpm2.h"
#include "tpm2_alg_util.h"
#include "tpm2_convert.h"
#include "tpm2_nv_util.h"
#include "tpm2_tool.h"

typedef struct tpm_setprimarypolicy_ctx tpm_setprimarypolicy_ctx;
struct tpm_setprimarypolicy_ctx {
    //Input
    struct {
        const char *ctx_path;
        const char *auth_str;
        tpm2_loaded_object object;
    } hierarchy;

    const char *policy_path;

    TPMI_ALG_HASH hash_algorithm;

    char *cp_hash_path;
};

static tpm_setprimarypolicy_ctx ctx = {
    .hash_algorithm = TPM2_ALG_NULL,
};

static bool set_digest_algorithm(char *value) {

    ctx.hash_algorithm = tpm2_alg_util_from_optarg(value, tpm2_alg_util_flags_hash);
    if (ctx.hash_algorithm == TPM2_ALG_ERROR) {
        LOG_ERR("Could not convert to number or lookup algorithm, got: "
                "\"%s\"", value);
        return false;
    }
    return true;
}

static bool on_option(char key, char *value) {

    bool result = true;

    switch (key) {
    case 'C':
        ctx.hierarchy.ctx_path = value;
        break;
    case 'P':
        ctx.hierarchy.auth_str = value;
        break;
    case 'L':
        ctx.policy_path = value;
        break;
    case 'g':
        result = set_digest_algorithm(value);
        break;
    case 0:
        ctx.cp_hash_path = value;
        break;
    }

    return result;
}

bool tpm2_tool_onstart(tpm2_options **opts) {

    static const struct option topts[] = {
        { "hierarchy",      required_argument, NULL, 'C' },
        { "auth",           required_argument, NULL, 'P' },
        { "policy",         required_argument, NULL, 'L' },
        { "hash-algorithm", required_argument, NULL, 'g' },
        { "cphash",         required_argument, NULL,  0  },
    };

    *opts = tpm2_options_new("C:P:L:g:", ARRAY_LEN(topts), topts,
        on_option, NULL, 0);

    return *opts != NULL;
}

static bool is_input_options_args_valid(void) {

    if (!ctx.hierarchy.ctx_path) {
        LOG_ERR("Must specify the hierarchy '-C'.");
        return false;
    }

    bool result = true;
    if (ctx.policy_path) {
        unsigned long file_size = 0;
        result = files_get_file_size_path(ctx.policy_path, &file_size);
        if (!result || file_size == 0) {
            result = false;
        }
    }

    if (ctx.cp_hash_path) {
        LOG_WARN("Calculating cpHash. Exiting without setting primary policy.");
    }

    return result;
}

static tool_rc process_setprimarypolicy_input(ESYS_CONTEXT *ectx,
    TPM2B_DIGEST **auth_policy) {

    /*
     * Load hierarchy handle and auth
     */
    tool_rc rc = tpm2_util_object_load_auth(ectx, ctx.hierarchy.ctx_path,
            ctx.hierarchy.auth_str, &ctx.hierarchy.object, false,
            TPM2_HANDLE_FLAGS_O|TPM2_HANDLE_FLAGS_P|TPM2_HANDLE_FLAGS_E|
            TPM2_HANDLE_FLAGS_L);
    if (rc != tool_rc_success) {
        return rc;
    }

    /*
     * Load policy digest if one is specified
     */
    if (ctx.policy_path) {

        *auth_policy = malloc(UINT16_MAX + sizeof(uint16_t));
        if (!*auth_policy) {
            LOG_ERR("oom");
            return tool_rc_general_error;
        }

        (*auth_policy)->size = UINT16_MAX;
        bool result = files_load_bytes_from_path(ctx.policy_path,
                (*auth_policy)->buffer, &((*auth_policy)->size));
        if (!result) {
            LOG_ERR("Failed loading policy digest from path");
            return tool_rc_general_error;
        }
    }

    return tool_rc_success;
}

tool_rc tpm2_tool_onrun(ESYS_CONTEXT *ectx, tpm2_option_flags flags) {

    /* opts is unused, avoid compiler warning */
    UNUSED(flags);

    bool result = is_input_options_args_valid();
    if (!result) {
        return tool_rc_option_error;
    }

    //Input
    TPM2B_DIGEST *auth_policy = NULL;
    tool_rc rc = process_setprimarypolicy_input(ectx, &auth_policy);
    if (rc != tool_rc_success) {
        return rc;
    }

    //ESAPI call
    if (ctx.cp_hash_path) {
        TPM2B_DIGEST cp_hash = { .size = 0 };
        rc = tpm2_setprimarypolicy(ectx, &ctx.hierarchy.object, auth_policy,
        ctx.hash_algorithm, &cp_hash);
        if (rc != tool_rc_success) {
            goto out;
        }

        result = files_save_digest(&cp_hash, ctx.cp_hash_path);
        if (!result) {
            rc = tool_rc_general_error;
        }
        goto out;
    }
    rc = tpm2_setprimarypolicy(ectx, &ctx.hierarchy.object, auth_policy,
        ctx.hash_algorithm, NULL);
out:
    free(auth_policy);
    return rc;
}

tool_rc tpm2_tool_onstop(ESYS_CONTEXT *ectx) {

    UNUSED(ectx);
    return tpm2_session_close(&ctx.hierarchy.object.session);
}
